Temporary experimental broker query API

korthout opened the pull request:
## Description

<!-- Please explain the changes you made here. -->

This adds a new query API to the broker. It is disabled by default. To enable it, set `zeebe.broker.network.commandapi.queryapi.enabled` to true. 

Usage requires you to write your own code to encode ExecuteQueryRequest and decode ErrorResponse and ExecuteQueryResponse messages. See `io.camunda.zeebe.broker.transport.externalapi.QueryApiIT` for an example of this. In particular, have a look at `io.camunda.zeebe.test.broker.protocol.queryapi.ExecuteQueryRequest` and `io.camunda.zeebe.test.broker.protocol.queryapi.ExecuteQueryResponse`.

At the moment of writing this pull request is not yet complete, additional tests and safe guards need to be added to the QueryApiRequestHandler (see the todo's in that class). In addition, request limiting is completely unimplemented. Lastly, changes to the configuration templates are still required to incorporate the new config option.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #7753
closes #7754

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

npepinpe commented on broker/src/test/java/io/camunda/zeebe/broker/transport/externalapi/QueryApiRequestHandlerTest.java:
```diff
@@ -0,0 +1,211 @@
+            final var buffer = new UnsafeBuffer(bytes);
+            serverResponse.write(buffer, 0);
+
+            messageHeaderDecoder.wrap(buffer, 0);
```
:wrench: Reuse the `ExecuteQueryResponse` test utility for wrapping it to simplify

npepinpe reviewed (changes requested):
:+1: :rocket: :fire: :100: 

- :x: Please move the `queryapi.enabled` flag into the experimental configuration section

Leftover todos:

- Implement query handling for JOB and PROCESS_INSTANCE (+ tests both unit and integration)
- Verify that the QueryService's ZeebeDb is closed only _after_ the partition listener has been called _and completed_, such that there is no wrong access to that ZeebeDb in the query handler after it has been closed
- Fix the race condition between query handling (in the transport actor) and the query service addition/removal (in the external API service actor); possibly make the request handler an actor as well?
- `commandApi` config parameter should be renamed to `externalApi` at some point in a backwards compatible way, i.e. if `commandApi` is defined then use that, otherwise use `externalApi` (possibly follow up issue, not required for initial iteration)

korthout commented:
@npepinpe I've rebased on develop and resolved the conflicts. All commits after the force-push are changes due to our pair review. Please have a look at those. I've resolved the corresponding conversations. 

I've also updated the pull request description, but here's an overview of everything that is still unfinished:

- [ ] resolve the existing todo's in the QueryApiRequestHandler
- [ ] move the config to experimental section
- [ ] verify that zeebedb closes _after_ the broker becomes follower (listener call)
- [ ] fix race condition between query handling (in the transport actor) and the query service addition/removal (in the external API service actor); possibly make the request handler an actor as well?
- [ ] rename `commandapi` config to `externalapi` in a backwards compatible way, can be postponed
- [ ] remove the open todo in QueryApiIT
- [ ] add `JOB` and `PROCESS_INSTANCE` cases to the QueryApiIT
- [ ] consider request limiting
- [ ] update config templates

Thanks again for taking over 🙇 and see you in 2 weeks 👋

npepinpe commented:
I'm going to summarize my changes here for Nico and anyone else who might look into it:

- Configuration is kept backwards compatible by adding a new `externalApi` field, but keeping the old `commandApi` field. So that you can now (and should) use `zeebe.broker.network.externalApi` for configuration, but if you still configure `zeebe.broker.network.commandApi`, this setting takes precedence and will overwrite anything you put in `externalApi`. This means we should update documentation in general. NOTE: we really should've made this a separate PR, as right now it feels a bit hidden here.
- Concurrent access when registering/closing the query service in the request handler is done by making the request handler an actor. As actors are our concurrency model, it seemed somewhat strange to use a command queue here like in the command API. I'm happy to challenge this however.
- Concurrent access to the `ZeebeDb` wrapped by the query service is managed via a volatile flag in the service. To set this, a new transition step was added, which will create (on open) and close (on close) the service. The listener is only called after the transition, so this is necessary to avoid accesses to a closed `ZeebeDb`, which produce seg faults. While having to do a volatile check on every call isn't great, since this API is temporary, it's fine to start like this. There were other options considered (e.g. wrapping the service in an actor), but they were all a little more complex (even though cleaner).
- Error messages were adapted to follower our usual pattern and be slightly more user friendly.

npepinpe reviewed (approved):
bors merge

npepinpe commented:
bors retry

npepinpe commented:
bors retry

npepinpe commented:
bors cancel

Actually, on second thought, I would rather keep the whole `externalapi` configuration part out for now. We can rename it internally (as we can then rename it as much as we want), but since changes to the configuration affects users, I'd like to think a little bit more about it before renaming it.

npepinpe commented:
bors retry

npepinpe commented:
bors retry

npepinpe commented:
bors retry

Sweet sweet flaky tests :cake:
